Speedup

tclausing opened the pull request:
I'm using this library in sort of a batch situation, where the user is sitting there waiting for it to finish, so I wanted to see if I could speed it up in some way. I tried two approaches - 
1) Since the end goal is having the DOM serialized to XHTML, then instead of actually cloning the elements, which is expensive, rather scan them and store the necessary information in a native object structure and serialize that.
2) Building on 1, use web workers to do the work in the middle of the pipeline (after the elements are scanned, and before the final Image is produced). Web workers are not allowed to touch the document or elements, so the goal was to get them out of the picture as early as possible and switch to the other thread.

I [benchmarked](https://codepen.io/anon/pen/MmORPO?editors=0011) each of these approaches along with the master branch against a fairly complicated DOM, in blocking and non-blocking flavors, where

Blocking is equivalent to
```
domtoimage.toPng(element).then(function() {
    domtoimage.toPng(element).then(function() {
        ...
    });
});
```

Non-blocking is equivalent to
```
Promise.all([
    domtoimage.toPng(element),
    domtoimage.toPng(element),
    ...
]).then(...);
```

Here are typical results I get from it on Chrome. "Original" refers to your master branch, and "Rewrite" and "RewriteWorkers" correspond to 1) and 2) above, respectively:
```
"Running tests..."
"blocking(singleThreaded(libs.domtoimageOriginal))"
"2.95 ops/sec"
"nonBlocking(singleThreaded(libs.domtoimageOriginal))"
"4.13 ops/sec"
"blocking(singleThreaded(libs.domtoimageRewrite))"
"4.71 ops/sec"
"nonBlocking(singleThreaded(libs.domtoimageRewrite))"
"9.16 ops/sec"
"blocking(multiThreaded(libs.domtoimageRewriteWorkers))"
"4.50 ops/sec"
"nonBlocking(multiThreaded(libs.domtoimageRewriteWorkers))"
"9.89 ops/sec"
```

Approach 1 gives a healthy improvement, 60% for blocking and 120% for non-blocking, but adding multithreading does not improve it any further. So it appears that actually cloning elements is relatively slow, while pulling the data off them and manipulating a native javascript object is quicker. Looking at CPU profiles of each one, it looks like multithreading doesn't help significantly because most of the work done in the middle of the pipeline is already non-blocking (XHR), so a single thread is able to perform nearly as well as a pool. The single thread is slightly faster in the blocking test since you're waiting for each iteration to finish, so there's no parallelism, only threading overhead. And it's slightly slower in the non-blocking test because the final serialization is a bit of blocking work that gives the threadpool a slight edge. However it's insignificant enough that this PR is for approach 1.

tclausing commented:
I added a new approach that builds on 1) and works well. It works better for a batch of different DOMs by offering a separate "scan" function that just collects the DOM information and returns it. The consumer can then pass that object into one of the toXYZ functions for final rendering, but they can also immediately move on to the next DOM and start scanning it; they don't have to wait for rendering of the first DOM to finish. So it still benefits nicely from non-blocking effects during the remainder of the pipeline, and it actually gets a speedup nearly as big as the full non-blocking test above, but it's more helpful in real world situations. In my real world situation I would be going from
```
"blocking(singleThreaded(libs.domtoimageOriginal))"
"2.95 ops/sec"
```
to
```
"chunk(libs.domtoimageRewriteChunk)"
"8.32 ops/sec"
```
for a nice 182% improvement.

sarath-sasikumar commented:
@tclausing , by how much is the approach 1 mentioned in the comment above improving the conversion performance.

tclausing commented:
@sarath-sasikumar for a single conversion (or for a sequence of blocking conversion calls), the improvement was 60%. You can run the benchmark yourself (on chrome) [here](https://codepen.io/anon/pen/MmORPO?editors=0011), comparing "blocking(singleThreaded(libs.domtoimageOriginal))" to 
"blocking(singleThreaded(libs.domtoimageRewrite))".

tsayen commented:
Hi. Sorry, I haven't had time to review those changes in depth yet (there is a lot :))
I'm afraid I do not fully understand the reasons why you cannot do it like in your 2nd example:
```
Promise.all([
    domtoimage.toPng(element1),
    domtoimage.toPng(element2),
    ...
]).then(...);
```
That example looks "asynchronous enough" to me actually. One might want to try to identify the hot spots in the code and think about moving them out of the main loop to the web worker, but that is not exactly the thing you're suggesting (please correct me if I'm wrong)

Could you please elaborate on why do we need the `scan()` function and all the complications that arise from that?

tclausing commented:
@tsayen Hello. The issue is that I have a batch of DOMs to render, and only one is loaded in the page at a time, so that prevents me from doing as you suggest above. I have to render one, trigger the next one to load, render that one, etc. So the scan function allows me to move on to the next DOM faster - it does only the "clone" parts of the pipeline, and then returns that intermediate data. I can now load the next DOM while simultaneously calling toPng() with the first one's scan() output (or using any other existing to* function). That's why I called this approach more helpful in real-world situations, outside of a simple benchmark. 

The "complication" was just to ensure that all steps that need the source DOM are executed during the scan/clone portion of the pipeline, so that the remaining steps can run without it. Scan consists of cloneNode and embedFonts, which you already had, and I added captureDimensions so that the options can be handled later without needing the DOM:

```javascript
    function scan(node, options) {
       options = options || {};
       return Promise.resolve({node: node})
           .then(function (ctx) {
               return cloneNode(ctx, options.filter, true);
           })
           .then(embedFonts)
           .then(captureDimensions);

           function captureDimensions(ctx) {
              ctx.nodeHeight = util.height(node);
              ctx.nodeWidth = util.width(node);
              return ctx;
           }
    }
```

Note that scan() is optional - toPng() can still be called directly as before, because toSvg() will internally call scan() if needed. Other than that, inlineBackground was also touching the source DOM, but the cloned/scanned object structure provides all the data it needs, so I switched it to use that instead of the DOM. The couple of places that were updating styling are now writing to a 'cssOverrides' array on the object structure that will ultimately be appended to the cssText string during serialization.

Note that even without taking advantage of scan(), all rendering functions get a speedup just from using a native object structure to store DOM info and serialize to xhtml, vs. cloning each element. Maybe I should have submitted two separate PR's...let me know if you like the native object approach but don't like scan(). And thanks for the fantasic library!

tamtakoe commented:
@tclausing did you check this changes in IE/Edge? Dom-to-img doesn't work there and returns error, but before this it blocks page UI. WebWorkers should solve this problem

tclausing commented:
@tamtakoe I don't recall, it's been a while. I'll do that in the next couple days... Though looks like I also need to resolve some conflicts from latest upstream to make sure I am truly isolating the issue. I've also been thinking I should split this PR into two, one to use simple objects to hold the relevant DOM properties instead of actually cloning the DOM; and second to decouple DOM scanning from image generation using web workers. It's possible that just number 1 would fix the problem (unless you already know otherwise?). Just need to find the time...

pie6k commented:
Would be great to have this merged!

pengw00 commented:
hello, is that Promiss.all speedup the process?
